feat(clis,pytest-plugins): allow pytest commands from package, add`execute eth-config` command

marioevz opened the pull request:
## 🗒️ Description

### Allow all pytest-plugin based commands to run outside of the repo

Allows all commands to run without having to clone the full repository. This was done by adding the pytest-*.ini config files to the pyproject.toml and also changing the working directory, root directory, and adding a new "working-directory" plugin to fix the working directory after pytest has been loaded, before the command runs.

### `execute eth-config`

Add `execute eth-config` command to test a running eth RPC client for the correct configuration given the specified network.

### `pytest.ini` -> `pytest-fill.ini`

The default pytest configuration file for `fill` has been renamed and moved to the appropriate folder, to include in the project build.

## 🔗 Related Issues or PRs
Closes #1885

## ✅ Checklist

- [x] All: Ran fast `tox` checks to avoid unnecessary CI fails, see also [Code Standards](https://eest.ethereum.org/main/getting_started/code_standards/) and [Enabling Pre-commit Checks](https://eest.ethereum.org/main/dev/precommit/):
    ```console
    uvx --with=tox-uv tox -e lint,typecheck,spellcheck,markdownlint
    ```
- [x] All: PR title adheres to the [repo standard](https://eest.ethereum.org/main/getting_started/contributing/?h=contri#commit-messages-issue-and-pr-titles) - it will be used as the squash commit message and should start `type(scope):`.
- [x] All: Considered adding an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [ ] All: Set appropriate labels for the changes (only maintainers can apply labels).

danceratopz commented on src/pytest_plugins/execute/eth_config/test_test_eth_config.py:
```diff
@@ -0,0 +1,141 @@
+"""Unit tests for the `eth_config` execute tests."""
```
I'd put this is in a `./tests/` sub-dir to match repo standards and avoid the `test_test` prefix :smile:

danceratopz commented on src/pytest_plugins/execute/eth_config/eth_config.py:
```diff
@@ -0,0 +1,48 @@
+from ethereum_test_forks import Fork
+from ethereum_test_rpc import EthRPC
+
+COMMAND_EXAMPLE = """
+uv run execute eth-config --chain-id 0x88bb0
+ --current Prague --current-activation-time 0 --next Osaka --next-activation-time 1742999832
+ --next-blob-schedule '{"baseFeeUpdateFraction": 5007716, "max": 9, "target": 6}'
+"""
```
This is unused and doesn't work?

danceratopz commented on src/pytest_plugins/execute/eth_config/eth_config.py:
```diff
@@ -0,0 +1,48 @@
+    eth_config_group.addoption(
+        "--network",
+        action="store",
+        dest="network",
```
Can we fail early if this value is invalid by checking it's availability in the available network configs in `pytest_configure()`?

danceratopz commented on src/ethereum_test_rpc/test_types.py:
```diff
@@ -0,0 +1,93 @@
+"""Test the types in the `ethereum_test_rpc` package."""
```
Shouldn't this be in a ./tests sub-folder?

danceratopz commented on src/pytest_plugins/execute/eth_config/networks.py:
```diff
@@ -0,0 +1,26 @@
+
+from .types import NetworkConfig
+
+NETWORK_CONFIGS = {
+    "Mainnet": NetworkConfig(
+        chain_id=1,
+        fork_activation_times={
+            Shanghai: 0,
+            Cancun: 0,
+        },
+    ),
+    "Hoodi": NetworkConfig(
+        chain_id=0x88BB0,
+        fork_activation_times={
+            Cancun: 0,
+            Prague: 1742999832,
+        },
+    ),
+}
```
Also:
```
  uv run execute eth-config \
      --network-config-file ./configs/l2-networks.json \
      --network arbitrum-sepolia \
      --rpc-endpoint http://arb-sepolia:8545
```

danceratopz commented on src/pytest_plugins/execute/eth_config/types.py:
```diff
@@ -0,0 +1,147 @@
+        return (
+            current_config.get_config(),
+            next_config.get_config() if next_config is not None else None,
+        )
```
Something like this to enable loading / saving network configs from / to JSON:
```suggestion
        )
        
  class NetworkConfigs(BaseModel):
      """Container for multiple network configurations."""

      networks: Dict[str, NetworkConfig]

      @classmethod
      def from_json_file(cls, file_path: str) -> "NetworkConfigs":
          """Load multiple network configs from JSON file."""
          with open(file_path) as f:
              return cls.model_validate_json(f.read())

      def to_json_file(self, file_path: str) -> None:
          """Save multiple network configs to JSON file."""
          with open(file_path, 'w') as f:
              f.write(self.model_dump_json(indent=2))

      def list_networks(self) -> List[str]:
          """List available network names."""
          return list(self.networks.keys())
```

danceratopz commented on src/pytest_plugins/execute/eth_config/types.py:
```diff
@@ -0,0 +1,147 @@
+                )
+                break
+        assert current_config is not None, f"Unable to determine current fork at {current_time}"
+        for bpo_activation_time in reversed(self.blob_parameter_only_forks.keys()):
```
The body of this loop is difficult to read.

danceratopz reviewed (commented):
Great idea to put this in `execute`! 

Love the integration with `src/ethereum_test_forks/forks/forks.py`.

Can we add a way to test this via hive? Does this need another sub-command,
`execute hive-eth-config`? This would allow us to verify before clients hit the devnet and help minimize disruption.

marioevz commented:
@danceratopz addressed all your comments and implemented this PR: https://github.com/ethereum/EIPs/pull/9989

Let me know what you think and thanks for the first review :)

danceratopz commented:
> LGTM, but the question still stands from the first review:
> 
> > Can we add a way to test this via hive? Does this need another sub-command,
> > `execute hive-eth-config`? This would allow us to verify before clients hit the devnet and help minimize disruption.

Thanks for explaining offline :) This doesn't make much sense to test in a staging environment like hive as we need to test the client with the network's genesis. In the EEST meeting today, we decided we can think of this as more of a health check that should be ran regularly on networks to ensure the clients running there meet the fork, and perhaps more relevantly, the next fork's spec.

marioevz commented on pytest-framework.ini:
```diff
@@ -17,3 +17,4 @@ addopts =
     --ignore=src/pytest_plugins/consume/simulators/engine/test_via_engine.py
     --ignore=src/pytest_plugins/consume/simulators/rlp/test_via_rlp.py
     --ignore=src/pytest_plugins/execute/test_recover.py
+    --ignore=src/pytest_plugins/execute/eth_config/test_eth_config.py
```
Changed `test_recover.py` too!

danceratopz commented on pyproject.toml:
```diff
@@ -115,7 +115,14 @@ exclude = ["tests/static/**"]
+    "direct/test_via_direct.py"
+]
 "cli.eest.make" = ["templates/*.j2"]
+"cli.pytest_commands" = ["pytest_ini_files/*.ini"]
```
Love it. Would have preferred a separate PR 😂 

Closes #1885

danceratopz replied:
Could you please move `fill`'s `pytest.ini` here too, required for #1885. But let's do `pytest-framework.ini` -> `pyproject.toml` in a follow-up.

danceratopz commented on pyproject.toml:
```diff
@@ -115,7 +115,14 @@ exclude = ["tests/static/**"]
 [tool.setuptools.package-data]
 ethereum_test_forks = ["forks/contracts/*.bin"]
+"pytest_plugins.execute" = ["eth_config/networks.yml"]
+"pytest_plugins.consume" = [
+    "simulators/hive_tests/__init__.py",
+    "simulators/hive_tests/test_via_*.py",
+    "direct/test_via_direct.py"
+]
```
I think this isn't necessary, as they are technically already in the package, no?

We don't have to include `src/pytest_plugins/execute/eth_config/execute_eth_config.py`, for example.

marioevz replied:
I just tried it out and it works, I'm not sure what led me to think this was necessary, it was in between the many trial-and-error attempts :P

Btw, the command I'm using to test this out is:
```bash
uv run --refresh --with /home/marioevz/Development/Eth/execution-spec-tests consume rlp --input=https://github.com/ethereum/execution-spec-tests/releases/download/v4.5.0/fixtures_stable.tar.gz
```
From a folder that does not contain eest code.

danceratopz replied:
`--refresh` from a local folder! Nice! Yeah, it's important to cd out of eest root :)

danceratopz commented on src/cli/pytest_commands/base.py:
```diff
@@ -34,22 +41,39 @@ def process_args(self, args: List[str]) -> List[str]:
         pass
 
 
+@contextmanager
```
TIL `@contextmanager`.

danceratopz commented on src/cli/pytest_commands/consume.py:
```diff
@@ -6,64 +6,51 @@
+    )
 
 
 def get_command_paths(command_name: str, is_hive: bool) -> List[Path]:
```
Thanks for the better variable naming! How about...?
```suggestion
def get_test_paths(command_name: str, is_hive: bool) -> List[Path]:
```

marioevz replied:
Used `get_static_test_paths`.

danceratopz commented on src/cli/pytest_commands/base.py:
```diff
@@ -1,21 +1,28 @@
 
+CURRENT_FOLDER = Path(realpath(__file__)).parent
+PYTEST_INI_FOLDER = CURRENT_FOLDER / "pytest_ini_files"
+ROOT_FOLDER = CURRENT_FOLDER.parent.parent
```
brittle but good, works anywhere!

danceratopz reviewed (changes requested):
Amazing! I was contemplating this yesterday as well, was perfect prep for this review 😄 
https://github.com/ethereum/execution-spec-tests/issues/1885

I also found `--rootdir` but hadn't put it all together - you've added a really nice solution!

I was wondering about putting the ini files next to the "entry" pytest plugin for the command, but they're clearly better in one place. Easier to diff; no searching required.
